src/autometrics/decorator.py

```diff
@@ -29,49 +36,102 @@ def autometrics(
+        )
+
+
+    def sync_decorator(func: Callable[P, T]) -> Callable[P, T]:
+        """Helper for decorating synchronous functions, to track calls and duration."""
 
-    def decorator(func: Callable[P, T]) -> Callable[P, T]:
         module_name = get_module_name(func)
         func_name = func.__name__
 
         @wraps(func)
-        def wrapper(*args: P.args, **kwds: P.kwargs) -> T:
+        def sync_wrapper(*args: P.args, **kwds: P.kwargs) -> T:
             start_time = time.time()
             caller = get_caller_function()
 
             try:
                 result = func(*args, **kwds)
-                get_tracker().finish(
+                track_result_ok(
+                    start_time, function=func_name, module=module_name, caller=caller
+                )
+
+            except Exception as exception:
+                result = exception.__class__.__name__
+                track_result_error(
                     start_time,
                     function=func_name,
                     module=module_name,
                     caller=caller,
-                    objective=objective,
-                    result=Result.OK,
+                )
+                # Reraise exception
+                raise exception
+            return result
+
+        sync_wrapper.__doc__ = append_docs_to_docstring(func, func_name, module_name)
+        return sync_wrapper
+
+
+    def async_decorator(func: Callable[P, Awaitable[T]]) -> Callable[P, Awaitable[T]]:
+        """Helper for decorating async functions, to track calls and duration."""
+
+        module_name = get_module_name(func)
+        func_name = func.__name__
+
+        @wraps(func)
+        async def async_wrapper(*args: P.args, **kwds: P.kwargs) -> T:
+            start_time = time.time()
+            caller = get_caller_function()
+
+            try:
+                result = await func(*args, **kwds)
+                track_result_ok(
+                    start_time, function=func_name, module=module_name, caller=caller
                 )
 
             except Exception as exception:
                 result = exception.__class__.__name__
-                get_tracker().finish(
+                track_result_error(
                     start_time,
                     function=func_name,
                     module=module_name,
                     caller=caller,
-                    objective=objective,
-                    result=Result.ERROR,
                 )
                 # Reraise exception
                 raise exception
             return result
 
-        if func.__doc__ is None:
-            wrapper.__doc__ = write_docs(func_name, module_name)
-        else:
-            wrapper.__doc__ = f"{func.__doc__}\n{write_docs(func_name, module_name)}"
-        return wrapper
+        async_wrapper.__doc__ = append_docs_to_docstring(func, func_name, module_name)
+        return async_wrapper
 
     if func is None:
-        return decorator
+        return sync_decorator
```

Comment:
i'm actually not quite sure what this accomplishes! now that we have two different decorators (one for sync, one for async) i am a slightly worried about returning the sync decorator here when there isn't a function. 

i did try a version of this PR that used one decorator, but to get the value of an async function synchronously, you have to do some tricks with the event loop which seemed like they were against best practices. e.g.,:

```py
import asyncio

async def async_function():
    # Your async code here
    return result

def sync_function():
    loop = asyncio.get_event_loop()
    result = loop.run_until_complete(async_function())
    return result
```

anyways, we don't seem to document this case (where the decorator is called without being attached to a function), so this might not be a problem

Reply:
Instead of returning different decorators, would it be possible to use only one and check whether the result is a coroutine? 
https://docs.python.org/3/library/inspect.html#inspect.iscoroutine